### PR TITLE
Inliner interface

### PR DESCRIPTION
**Pull request by gysit**

### CODE REVIEW

**Review by definelicht — COMMENTED** [review]
Looks good, as long as we can get the `LLVM::LLVMFuncOp`. We will still have the issue of missing the `CallableOpInterface` operation in `processInlinedBlocks` if we want to do more fancy stuff there, which sounds like it will be harder to fix than we thought, since the callee region is invalid at this point...

**Comment by gysit** [comment]
> Looks good, as long as we can get the `LLVM::LLVMFuncOp`. We will still have the issue of missing the `CallableOpInterface` operation in `processInlinedBlocks` if we want to do more fancy stuff there, which sounds like it will be harder to fix than we thought, since the callee region is invalid at this point...

Yes I think getting the `CallableOpInterface` in `processInlinedBlocks` is difficult since it is technically dead at this point. Do we really need it there? I thought it is sufficient if we have it in `handleCallSiteArguments` and `handleCallSiteResults`?

**Comment by definelicht** [comment]
> Yes I think getting the CallableOpInterface in processInlinedBlocks is difficult since it is technically dead at this point. Do we really need it there? I thought it is sufficient if we have it in handleCallSiteArguments and handleCallSiteResults?

Maybe we don't need it for anything else 🤷‍♂️ So we can deal with it when it happens

File: mlir/include/mlir/Transforms/InliningUtils.h [hunk 1]
```diff
@@ -141,6 +142,41 @@ class DialectInlinerInterface
...
+  /// conversion result or the 'argument' itself if a conversion was not needed.
+  /// The type of the returned values has to match 'targetType' and the
+  /// 'argumentAttrs' dictionary is non-null even if no attribute is present.
+  /// The hook is called after converting the callsite argument types using the
+  /// materializeCallConversion callback and right before inlining the callee
+  /// region. Any operations created using the provided 'builder' are inserted
+  /// right before the inlined callee region. Example use cases are the
+  /// insertion of copies for by value arguments or integer conversions that
+  /// require signedness information.
+  virtual Value handleArgumentConversion(OpBuilder &builder, Operation *call,
```
Comment (definelicht) [inline mlir/include/mlir/Transforms/InliningUtils.h:155]: Why `Conversion`? Isn't this more general than that?

File: mlir/include/mlir/Transforms/InliningUtils.h [hunk 2]
```diff
@@ -141,6 +142,41 @@ class DialectInlinerInterface
...
+  /// Hook to handle attribute based argument conversions that return the
+  /// conversion result or the 'argument' itself if a conversion was not needed.
+  /// The type of the returned values has to match 'targetType' and the
+  /// 'argumentAttrs' dictionary is non-null even if no attribute is present.
+  /// The hook is called after converting the callsite argument types using the
+  /// materializeCallConversion callback and right before inlining the callee
+  /// region. Any operations created using the provided 'builder' are inserted
+  /// right before the inlined callee region. Example use cases are the
+  /// insertion of copies for by value arguments or integer conversions that
+  /// require signedness information.
```
Comment (definelicht) [inline mlir/include/mlir/Transforms/InliningUtils.h:154]:
```suggestion
  /// Hook to handle attribute-based argument conversions, returning the
  /// conversion result or the `argument` itself if a conversion was not needed.
  /// The type of the returned value has to match 'targetType', and the
  /// 'argumentAttrs' dictionary is non-null even if no attribute is present.
  /// The hook is called after converting the callsite argument types using the
  /// materializeCallConversion callback, and right before inlining the callee
  /// region. Any operations created using the provided `builder` are inserted
  /// right before the inlined callee region. Example use cases are the
  /// insertion of copies for by value arguments, or integer conversions that
  /// require signedness information.
```

File: mlir/include/mlir/Transforms/InliningUtils.h [hunk 3]
```diff
@@ -141,6 +142,41 @@ class DialectInlinerInterface
...
+  /// conversion result or the 'result' itself if a conversion was not needed.
+  /// The type of the returned values has to match 'targetType' and the
+  /// 'resultAttrs' dictionary is non-null even if no attribute is present. The
+  /// hook is called right before handling the terminators and obtains the
+  /// callee result before converting its type using the
+  /// materializeCallConversion callback. Any operations created using the
+  /// provided 'builder' are inserted right after the inlined callee region.
+  /// Example use cases are the insertion of copies for by value results or
+  /// integer conversions that require signedness information.
+  /// NOTE: This hook is invoked after inlining the 'callable' region.
```
Comment (definelicht) [inline mlir/include/mlir/Transforms/InliningUtils.h:172]:
```suggestion
  /// Hook to handle attribute-based result conversion, returning the
  /// conversion result or the `result` itself if a conversion was not needed.
  /// The type of the returned values has to match `targetType`, and the
  /// `resultAttrs` dictionary is non-null even if no attribute is present. The
  /// hook is called right before handling terminators, and obtains the
  /// callee result before converting its type using the
  /// `materializeCallConversion` callback. Any operations created using the
  /// provided `builder` are inserted right after the inlined callee region.
  /// Example use cases are the insertion of copies for by value results or
  /// integer conversions that require signedness information.
  /// NOTE: This hook is invoked after inlining the 'callable' region.
```

File: mlir/include/mlir/Transforms/InliningUtils.h [hunk 4]
```diff
@@ -141,6 +142,41 @@ class DialectInlinerInterface
...
+  /// The type of the returned values has to match 'targetType' and the
+  /// 'argumentAttrs' dictionary is non-null even if no attribute is present.
+  /// The hook is called after converting the callsite argument types using the
+  /// materializeCallConversion callback and right before inlining the callee
+  /// region. Any operations created using the provided 'builder' are inserted
+  /// right before the inlined callee region. Example use cases are the
+  /// insertion of copies for by value arguments or integer conversions that
+  /// require signedness information.
+  virtual Value handleArgumentConversion(OpBuilder &builder, Operation *call,
+                                         Operation *callable, Value argument,
```
Comment (definelicht) [inline mlir/include/mlir/Transforms/InliningUtils.h:156]: Can we really not include `CallOpInterface` and `CallableOpInterface`?

**Review by definelicht — COMMENTED** [review]
Let's see how they react 🙂

File: mlir/include/mlir/Transforms/InliningUtils.h (hunk 4 above)
Comment (gysit) [reply mlir/include/mlir/Transforms/InliningUtils.h:156]: Let's not do this. Most if not all of the stuff exposed on the interfaces is anyways already accessible via the other hook arguments. If we use the call or the callable then most likely anyways to read some attribute that is specific to the dialect and not exposed on the interface. I would thus say following the other hooks makes sense here.

File: mlir/include/mlir/Transforms/InliningUtils.h (hunk 3 above)
Comment (gysit) [reply mlir/include/mlir/Transforms/InliningUtils.h:172]: nice tnx

**Review by definelicht — APPROVED** [review]
Very curious to see what happens 😛

**Comment by gysit** [comment]
most likely not a lot :)

**Pull request closed without merging**